CI: bugtool and associated debugging information not gathered in failed builds

In https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3259/ , a test failed: 

```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:338
Expected
    <*errors.errorString | 0xc4202a9130>: {
        s: "Timeout reached: could not import policy",
    }
to be nil
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/runtime/monitor.go:116
```

But no logs were gathered; attached is the zipfile containing all gathered logs for this build.

Logs: [test_results_Cilium-PR-Ginkgo-Tests-Validated_3259.zip](https://github.com/cilium/cilium/files/2010768/test_results_Cilium-PR-Ginkgo-Tests-Validated_3259.zip)

In `RuntimeValidatedMonitorTestWithSampleContainersCiliummonitoreventtypes` there are only two files:

* cmds.log
* test-output.log

For some reason, the bugtool output is not contained here. And, there is no information in the Jenkins console output indicating what happened, nor in `test-output.log`.

Confirmed that this is indeed the issue.

Reproducer below. I did this for the K8s VMs because I already had them running on my local setup, this could be done with the Docker VMs as well most likely:

1. Boot the vms:

```
$ cd test/
$ vagrant up k8s1-1.10
...
$ vagrant up k8s2-1.10
...
```
2. Apply the following diff (this is done off of latest master):
```
diff --git a/test/helpers/kubectl.go b/test/helpers/kubectl.go
index 9cfc5d95c..01961db47 100644
--- a/test/helpers/kubectl.go
+++ b/test/helpers/kubectl.go
@@ -1003,7 +1003,7 @@ func (kub *Kubectl) ValidateNoErrorsOnLogs(duration time.Duration) {
        }
        logs := res.Output().String()
        for _, message := range checkLogsMessages {
-               gomega.ExpectWithOffset(1, logs).ToNot(gomega.ContainSubstring(message),
+               gomega.ExpectWithOffset(1, logs).To(gomega.ContainSubstring(message),
                        "Found a %q in Cilium logs", message)
        }
 }
```

3. Run the test without provisioning Cilium (this provides for fast turnaround time for testing):

```K8S_VERSION=1.10 ginkgo --focus="K8sValidatedHealthTest" -v -failFast=true -- -cilium.provision=false```

4. Test should fail like below:
```
Summarizing 1 Failure:

[Fail] K8sValidatedHealthTest [AfterEach] checks cilium-health status between nodes 
/Users/ianvernon/go/src/github.com/cilium/cilium/test/k8sT/Health.go:51

Ran 1 of 285 Specs in 38.635 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 284 Skipped --- FAIL: TestTest (38.65s)
FAIL
```

5. Test results directory contains no bugtool, other debugging output apart from files which are written to during test:

```
$ ls test_results/110-/K8sValidatedHealthTestcheckscilium-healthstatusbetweennodes/
cmds.log	test-output.log
```

Assigning to @eloycoto for fixing.

--- a/test/ginkgo-ext/scopes.go
+++ b/test/ginkgo-ext/scopes.go
@@ -51,6 +51,15 @@ var (
 	rootScope           = currentScope
 	countersInitialized bool
 
+	// failEnabled for tests that have failed on JustAfterEach function we need
+	// to handle differently, because `ginkgo.Fail` do a panic, and all the
+	// following functions will not be called. With the WrapFailfn if the fail
+	// is on any After function, will not panic, will mark the test as failed,
+	// and will trigger the Fail function at the end.
+	failEnabled     = true
+	afterEachFailed = map[string]bool{}
+	afterEachCB     = map[string]func(){}
+
 	Context                               = wrapContextFunc(ginkgo.Context, false)
 	FContext                              = wrapContextFunc(ginkgo.FContext, true)
 	PContext                              = wrapNilContextFunc(ginkgo.PContext)
@@ -68,7 +77,7 @@ var (
 	BeforeSuite                           = ginkgo.BeforeSuite
 	AfterSuite                            = ginkgo.AfterSuite
 	Skip                                  = ginkgo.Skip
-	Fail                                  = ginkgo.Fail
+	Fail                                  = FailWithToggle
 	CurrentGinkgoTestDescription          = ginkgo.CurrentGinkgoTestDescription
 	GinkgoRecover                         = ginkgo.GinkgoRecover
 	GinkgoT                               = ginkgo.GinkgoT
@@ -219,8 +228,9 @@ func runAllAfterFail(cs *scope, testName string) {
 		return
 	}
 
+	hasFailed, _ := afterEachFailed[testName]
 	for _, body := range cs.afterFail {
-		if ginkgo.CurrentGinkgoTestDescription().Failed {
+		if ginkgo.CurrentGinkgoTestDescription().Failed || hasFailed {
 			body()
 		}
 	}
@@ -236,7 +246,20 @@ func RunAfterEach(cs *scope) {
 	if cs == nil {
 		return
 	}
+
+	// Disabling the `ginkgo.Fail` function to avoid the panic and be able to
+	// gather all the logs.
+	failEnabled = false
+	defer func() {
+		failEnabled = true
+	}()
+
 	testName := ginkgo.CurrentGinkgoTestDescription().FullTestText
+
+	if _, ok := afterEachFailed[testName]; !ok {
+		afterEachFailed[testName] = false
+	}
+
 	runAllJustAfterEach(cs, testName)
 	justAfterEachStatus[testName] = true
 
@@ -256,6 +279,11 @@ func RunAfterEach(cs *scope) {
 		}
 	}
 	after()
+
+	cb := afterEachCB[testName]
+	if cb != nil {
+		cb()
+	}
 }
 
 // AfterEach runs the function after each test in context
@@ -414,4 +442,19 @@ func calculateCounters(s *scope, focusedOnly bool) (int, bool) {
 	}
 	s.counter = int32(count)
 	return count, haveFocused
+}
+
+// FailWithToggle wraps `ginkgo.Fail` function to have a option to disable the
+// panic when something fails when is running on AfterEach.
+func FailWithToggle(message string, callerSkip ...int) {
+	if failEnabled {
+		ginkgo.Fail(message, callerSkip...)
+	}
+
+	testName := ginkgo.CurrentGinkgoTestDescription().FullTestText
+	afterEachFailed[testName] = true
+
+	afterEachCB[testName] = func() {
+		ginkgo.Fail(message, callerSkip...)
+	}
 }

--- a/test/helpers/utils.go
+++ b/test/helpers/utils.go
@@ -197,7 +197,7 @@ func InstallExampleCilium(kubectl *Kubectl) {
 		KubeSystemNamespace, "-l k8s-app=cilium", timeout)
 	ExpectWithOffset(1, err).Should(BeNil(), "Cilium is not ready after timeout")
 
-	ginkgo.By(fmt.Sprintf("Checking that installed image is %q", StableImage))
+	ginkgoext.By(fmt.Sprintf("Checking that installed image is %q", StableImage))
 
 	filter := `{.items[*].status.containerStatuses[0].image}`
 	data, err := kubectl.GetPods(
@@ -290,7 +290,7 @@ func Fail(description string, callerSkip ...int) {
 		fmt.Fprintf(os.Stdout, "\nRun \"kill -SIGCONT %d\" to continue.\n", pid)
 		syscall.Kill(pid, syscall.SIGSTOP)
 	}
-	ginkgo.Fail(description, callerSkip...)
+	ginkgoext.Fail(description, callerSkip...)
 }
 
 // CreateReportDirectory creates and returns the directory path to export all report

